substrate/frame/assets/src/functions.rs

```diff
@@ -228,20 +247,21 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 		let account = Account::<T, I>::get(&id, who).ok_or(Error::<T, I>::NoAccount)?;
 		ensure!(!account.status.is_frozen(), Error::<T, I>::Frozen);
 
-		let amount = if let Some(frozen) = T::Freezer::frozen_balance(id, who) {
-			// Frozen balance: account CANNOT be deleted
-			let required =
-				frozen.checked_add(&details.min_balance).ok_or(ArithmeticError::Overflow)?;
-			account.balance.saturating_sub(required)
-		} else {
-			if keep_alive {
-				// We want to keep the account around.
-				account.balance.saturating_sub(details.min_balance)
-			} else {
-				// Don't care if the account dies
-				account.balance
-			}
+		let untouchable = match (
+			T::Holder::balance_on_hold(id.clone(), who),
+			T::Freezer::frozen_balance(id.clone(), who),
+			keep_alive,
+		) {
+			(None, None, true) => details.min_balance,
+			(None, None, false) => Zero::zero(),
+			(maybe_held, maybe_frozen, _) => {
+				let held = maybe_held.unwrap_or_default();
+				let frozen = maybe_frozen.unwrap_or_default();
+				frozen.saturating_sub(held).max(details.min_balance)
+			},
 		};
+		let amount = account.balance.saturating_sub(untouchable);
```

Comment:
let's document `balance` field better. otherwise the logic for obtaining the `untouchable` wont be clear for a new reader. we need to communicate that this `balance` includes the frozen balance, if any from `Config::Freezer`. And does not include the hold balance, if any from `Config::Holder`.

Reply:
Done on [`6bebe7e`](https://github.com/paritytech/polkadot-sdk/pull/4530/commits/6bebe7e403774ba3565cb54ecee3d7d10ad788d5)